SofaKernel/modules/SofaCore/src/sofa/core/DataTracker.cpp

```diff
@@ -92,17 +93,24 @@ void DataTrackerDDGNode::updateAllInputsIfDirty()
     }
 }
 ///////////////////////
-void DataTrackerEngine::addCallback( std::function<void(DataTrackerEngine*)> f)
+void DataTrackerEngine::addCallback( std::function<sofa::core::objectmodel::ComponentState(void)> f)
 {
     m_callbacks.push_back(f);
 }
 
 void DataTrackerEngine::update()
 {
+    updateAllInputsIfDirty();
+    sofa::core::objectmodel::ComponentState cs = sofa::core::objectmodel::ComponentState::Valid;
     for(auto& callback : m_callbacks)
     {
-        callback(this);
+        sofa::core::objectmodel::ComponentState state = callback();
+        if (state != sofa::core::objectmodel::ComponentState::Valid)
+            cs = state;
     }
+    if (m_owner)
+        m_owner->d_componentstate.setValue(cs);
+    cleanDirty();
 }
```

Comment:
The reason why the state is valid by default is that if there's no callback (which is the case most often) we can't assume that the state is invalid: 99.X% of components in sofa aren't using callbacks right now, so seeing their state to invalid would set every component's state to invalid as soon as one of their data change. If the system was in place since the early days of sofa, setting it to invalid by default would enforce the use of the mechanism and be the right choice, but we have to deal with how things currently are in sofa

Reply:
It is a non trivial part. I have no opinion on what is the best appraoch

Reply:
Solution to the problem I chose:
Remove the use of callback engines accepting more than 1 callback function. There isn't any use case that I know about where multiple callback functions on the same set of inputs / outputs make sense AFAIK. Base still possesses a list of callback engines, so multiple callback functions can be set per component. But within a callback engine, there is only one single function, thus no question regarding which state should predominate.

Note that when creating a component with multiple callbacks, the returned componentState should take into account ALL callback results. Take this example:

A component has 2 inputs (in1, in2) and 2 outputs (out1, out2), connected respectively using callback1 and callback2.
- in1 contains a value that doesn't allow for out1 to contain valid information, thus in callback1, the returned state is Invalid. 
- in2 contains an invalid value, so out2 contains invalid information. thus in callback2, the returned state is Invalid
-> everything is OK, the componentState is properly set to Invalid.

Someone changes in2 to a valid value, the callback is called, and out2 is updated and now contains valid info. Thus callback2 returns a Valid status
-> Problem because out1 is still invalid. Only 1 invalid output is enough to set the whole component's state to invalid.

CCL: callback1 and callback2 should both check for the validity of ALL outputs before deciding whether the componentState should be valid or not. Kind of obvious but important to note